symbol.c

```diff
@@ -82,23 +93,301 @@ enum id_entry_type {
     ID_ENTRY_SIZE
 };
 
+typedef struct {
+    rb_id_serial_t last_id;
+    VALUE sym_set;
+
+    VALUE ids;
+} rb_symbols_t;
+
 rb_symbols_t ruby_global_symbols = {tNEXT_ID-1};
 
-static const struct st_hash_type symhash = {
-    rb_str_hash_cmp,
-    rb_str_hash,
+struct sym_set_static_sym_entry {
+    VALUE sym;
+    VALUE str;
+};
+
+#define SYM_SET_SYM_STATIC_TAG 1
+
+static bool
+sym_set_sym_static_p(VALUE sym)
+{
+    return sym & SYM_SET_SYM_STATIC_TAG;
+}
+
+static VALUE
+sym_set_static_sym_tag(struct sym_set_static_sym_entry *sym)
+{
+    VALUE value = (VALUE)sym | SYM_SET_SYM_STATIC_TAG;
+    RUBY_ASSERT(IMMEDIATE_P(value));
+    RUBY_ASSERT(sym_set_sym_static_p(value));
+
+    return value;
+}
+
+static struct sym_set_static_sym_entry *
+sym_set_static_sym_untag(VALUE sym)
+{
+    RUBY_ASSERT(sym_set_sym_static_p(sym));
+
+    return (struct sym_set_static_sym_entry *)(sym & ~((VALUE)SYM_SET_SYM_STATIC_TAG));
+}
+
+static VALUE
+sym_set_sym_get_str(VALUE sym)
+{
+    VALUE str;
+    if (sym_set_sym_static_p(sym)) {
+        str = sym_set_static_sym_untag(sym)->str;
+    }
+    else {
+        RUBY_ASSERT(RB_TYPE_P(sym, T_SYMBOL));
+        str = RSYMBOL(sym)->fstr;
+    }
+
+    RUBY_ASSERT(RB_TYPE_P(str, T_STRING));
+
+    return str;
+}
+
+static VALUE
+sym_set_hash(VALUE sym)
+{
+    if (sym_set_sym_static_p(sym)) {
+        return (VALUE)rb_str_hash(sym_set_static_sym_untag(sym)->str);
+    }
+    else {
+        return (VALUE)RSYMBOL(sym)->hashval;
+    }
+}
+
+static bool
+sym_set_cmp(VALUE a, VALUE b)
+{
+    return rb_str_hash_cmp(sym_set_sym_get_str(a), sym_set_sym_get_str(b)) == false;
+}
+
+
+static int
+sym_check_asciionly(VALUE str, bool fake_str)
+{
+    if (!rb_enc_asciicompat(rb_enc_get(str))) return FALSE;
+    switch (rb_enc_str_coderange(str)) {
+      case ENC_CODERANGE_BROKEN:
+        if (fake_str) {
+            str = rb_enc_str_new(RSTRING_PTR(str), RSTRING_LEN(str), rb_enc_get(str));
+        }
+        rb_raise(rb_eEncodingError, "invalid symbol in encoding %s :%+"PRIsVALUE,
+                 rb_enc_name(rb_enc_get(str)), str);
+      case ENC_CODERANGE_7BIT:
+        return TRUE;
+    }
+    return FALSE;
+}
+
+static VALUE
+dup_string_for_create(VALUE str)
+{
+    rb_encoding *enc = rb_enc_get(str);
+
+    str = rb_enc_str_new(RSTRING_PTR(str), RSTRING_LEN(str), enc);
+
+    rb_encoding *ascii = rb_usascii_encoding();
+    if (enc != ascii && sym_check_asciionly(str, false)) {
+        rb_enc_associate(str, ascii);
+    }
+    OBJ_FREEZE(str);
+
+    str = rb_fstring(str);
```

Comment:
The code was already like that, but after you merge I might followup to use encindex instead of `rb_encoding *` here, because `rb_enc_get` and other functions that deal with `rb_encoding *` are way slower.